[Wang Jun Hong] iP

WJunHong opened the pull request:
# ChiBot
> "The world isn't perfect. But it's there for us, doing the best it can....that's what makes it so damn beautiful." - Roy Mustang

ChiBot is your own personal assistant in managing your everyday tasks. She is:
- Text-based
- ~~Easy~~ Effortless to learn
- Super  *friendly* 

To start using ChiBot, all you need to do is:
1. Download it [here](https://github.com/WJunHong/ip)
2. Double-click it (or do a `git fork`)
3. Add your tasks
4. Put your mind at ease

Best of all, it's **FREE** 😸 

Some cool things ChiBot can do:
- [x] Add a new todo, deadline, or event!
- [ ] Check or uncheck tasks at will!
- [ ] Delete tasks you have completed!

To do basic version control with ChiBot, here are some useful git commands:
```
git status
git add file1 file2
git commit -m Enter your commit message
git push origin master
```
To run ChiBot manually from your favorite IDE, click into *Chi.java* file and run the following method:
```
    public static void main(String[] args) {
        Chi myBot = new Chi("data/tasks.txt");
        myBot.run();
    }
```

So what are you waiting for, Chi-san is waiting for your next request!

Commit:
Add Gradle support

WJunHong committed:
Added initial functionality to echo messages sent by user

WJunHong committed:
Added functionality to store and print messages

WJunHong committed:
added new Task class

WJunHong committed:
Added status message and simple mark and unmark features

WJunHong committed:
Added new subclasses of Task

WJunHong committed:
Added toString() for Tasks and Chi's todo handler

WJunHong committed:
Added response to Event and Deadline, changed event text

WJunHong committed:
Fixed styling issue, added some input cases and expected output

WJunHong committed:
Added more test cases, and updated javadocs for classes.

WJunHong committed:
Added exception class and started migrating code to function

WJunHong committed:
Added custom error messages for different tasks, added more handlers for tasks

WJunHong committed:
Added some comments explaining function

WJunHong committed:
Added and tested more cases with the new functionality, and added javadocs to Duke

WJunHong committed:
Added test case for empty task and date.

WJunHong committed:
Updated styling, javadocs and plans for the future.

WJunHong committed:
Added delete functionality, and exception handling for delete. Added more test cases for delete

WJunHong committed:
Add new class Main for driver function, rename files to match bot's name

WJunHong committed:
Implement a function to check and create new data file

WJunHong committed:
Add constants for data file and folder path, and shifted data storage creation into file parser function

WJunHong committed:
Add storage and load functionality to Chi bot

WJunHong committed:
Change string to datetime objects in deadline and event, add date and time check for parser

WJunHong committed:
Add custom text for date and times for event and deadlines

WJunHong committed:
Merge branch-Level-7 and branch-Level-8, integrate features of both into files

WJunHong committed:
Create new classes to abstract functionality away from Chi class

WJunHong committed:
Shift code to various components, and refactor code to fit new classes

WJunHong committed:
Remove obsolete methods and added more text checking for parser

WJunHong committed:
Implement the processing of deadline and event messages and correct some problems with how events and deadlines are stored and retrieved from file

WJunHong committed:
Add classes to package

WJunHong committed:
Add packages for each type of class

WJunHong committed:
Test parser and tasklist classes, add stubs for tasklist, task and storage

WJunHong committed:
Create jar file for project distribution

WJunHong committed:
Remove redundant files and commented out code

WJunHong committed:
Add javadoc for chi and exception classes

WJunHong committed:
Add javadocs for parser

WJunHong committed:
Add storage javadocs

WJunHong committed:
Add javadoc for tasks and changed spacing between javadoc description and param tag

WJunHong committed:
Add javadocs for todo

WJunHong committed:
Add javadocs for tasklist, taskliststub and ui components

WJunHong committed:
Fix indentation for try block and case statements for parser

WJunHong committed:
Fix wrapping indentation and switch statement for storage class

WJunHong committed:
Fix the indentation and spacing for task classes

WJunHong committed:
Uncomment ui style

WJunHong committed:
Fix name for delete task test case

WJunHong committed:
Add functions to handle finding keywords in parser, task and tasklist

WJunHong committed:
Fix a newline problem when storing data in file

WJunHong committed:
Resolve merge conflicts between javadocs and coding standard branches

WJunHong committed:
Resolve merge conflict between coding standards and find function

WJunHong committed:
Set default time if time is not indicated in deadline or event

WJunHong committed:
Merge branch 'add-gradle-support'

WJunHong committed:
Start Gradle config

WJunHong committed:
Update gradle build file

WJunHong committed:
Add checkstyle to project

WJunHong committed:
Add gradle dependencies for GUI

WJunHong committed:
Add some starter files for GUI creation

WJunHong committed:
Add parser functionality to GUI, remove checkstyle for gradle build

WJunHong committed:
Add command package to deal with everything command related

WJunHong committed:
Add some data to data file

WJunHong committed:
Implement exit functionality for JavaFX

WJunHong committed:
Add javadocs for new command classes

WJunHong committed:
Add more javadocs

WJunHong committed:
Add javadocs for JavaFX components

WJunHong committed:
Test delete functionality with GUI

MechFroG88 commented on src/main/java/duke/Main.java:
```diff
@@ -0,0 +1,24 @@
+package duke;
```
Your chatbot has a nice name!

Perhaps you can consider changing the package name from Duke to Chi? 🤔

MechFroG88 commented on src/main/java/duke/commands/Command.java:
```diff
@@ -0,0 +1,51 @@
+     */
+    public static Command of(Keywords keyword, String[] tokens) throws ChiException {
+        switch(keyword.toString()) {
+        case "ADD" :
+            return new AddCommand(tokens);
+        case "DELETE" :
+            return new DeleteCommand(Arrays.copyOfRange(tokens, 1, tokens.length ));
+        case "MARK" :
```
Maybe you can make the switch cases look more consistent by removing the space before `:`

WJunHong replied:
Haha didn't notice the space, thanks for pointing it out!

MechFroG88 commented on src/main/java/duke/commands/AddCommand.java:
```diff
@@ -0,0 +1,227 @@
+            switch(command) {
+            case "todo":
+                if (validateMessageBody(this.description, command)) {
+                    throw new ChiException("This todo has some problems nyan!");
```
Hahahah the `nyan` is so cute!

MechFroG88 commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,157 @@
+        } else {
+            if (splitTask.length == 6) {
+                return new Event(splitTask[2], LocalDate.parse(splitTask[3], DateTimeFormatter.ofPattern("yyyy-MM-dd")),
+                        LocalTime.parse(splitTask[4], DateTimeFormatter.ofPattern("HH:mm")),
+                        LocalTime.parse(splitTask[5], DateTimeFormatter.ofPattern("HH:mm")),
```
I think "HH:mm" can be set as a constant since it is repeated multiple times.

WJunHong replied:
Yup, thanks for the recommendation!

MechFroG88 reviewed (commented):
Overall really nice code!

Just a few suggestion that maybe you can look into.

WJunHong committed:
Credit code to solution by classmate

brian16600 commented on src/main/java/duke/commands/AddCommand.java:
```diff
@@ -0,0 +1,227 @@
+ */
+public class AddCommand extends Command {
+
+    /** The type of add command */
```
Good naming of variables - easy to understand!

brian16600 commented on src/main/java/duke/commands/AddCommand.java:
```diff
@@ -0,0 +1,227 @@
+    /**
+     * Obtains the task description in the message.
+     *
+     * @param msg The message sent.
```
Perhaps the naming could be clearer, such as commandType?

brian16600 commented on src/main/java/duke/commands/Command.java:
```diff
@@ -0,0 +1,55 @@
+     * @param tokens The entire content of the user message.
+     * @return A new command of the keyword type.
+     * @throws ChiException If are some problems in creating the command.
+     */
```
The switch/case section here is very simple and easy to read!

brian16600 commented on src/main/java/duke/commands/DeleteCommand.java:
```diff
@@ -0,0 +1,86 @@
+            if (words.length > 1) {
+                return false;
+            }
+            int s = Integer.parseInt(words[0]);
```
Perhaps the variable naming of s could be clearer here?

brian16600 commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,157 @@
+                        LocalTime.parse(splitTask[4], DateTimeFormatter.ofPattern("HH:mm")),
+                        splitTask[1].equals("1"));
+            }
+            /*else {
```
Maybe it might be better to remove the commented out else statement if it's not in use anymore?

brian16600 reviewed (commented):
Overall incredibly well done and a pleasure to read! The code was well organised and documented and java standards/conventions were well followed.

WJunHong committed:
Made changes to style based on checkstyle violations

WJunHong committed:
Change mainClassName for build.gradle jar file generation

Previously, the reference was made to Chi class. However, since
implementing JavaFX, the entry point is now at the Launcher class.

WJunHong committed:
Add help command feature's initial stages

Users might need some help using the application, especially for
new users.

Hence, with the help command, users can help troubleshoot
when Chi is unable to process their command.

The help command will ideally contain 2 main functionality, a general
help function and to help a specific command.
* help - Lists all possible help options
* help <command> - List how <command> should be formatted for Chi to
understand.

WJunHong committed:
Add assertions

It is important to check that the number of tasks in the list, after
adding one, is not 0. This verifies that the adding function to the
task list is working fine.

Furthermore, we need to check if the index we specified is based on
conventional array-based indexing and not the indexing used by the
user.

Lastly, we also need to check that tasks are stored properly so that
there will not be any parsing problems later on.

WJunHong committed:
Remove some redundant comments

Previously, there were comments repeating the code.

Now, some of those comments are beginning to be removed.

WJunHong committed:
Refactor add command

Many convoluted boolean expressions in the if block.

Broke expressions down to named variables to make sense of what we are
checking in the if block.

WJunHong committed:
Made fields final and reduce magic numbers

Previously, fields which would not change after assignment were not
set to final. Also, there were some magic numbers regarding command
length.

Now, fields are set to final to verify they will not be changed. Also,
magic numbers are given meanings.

WJunHong committed:
Remove dead code, make method names more meaningful

Previously, there was dead code from past task constructors and
scrapped implementation. Also, file writing method name was not
descriptive enough.

Now, dead code is reduced, and the method has a more meaningful name.

WJunHong committed:
Change package name

Previously, the package name was still duke, didn't make any sense.

Now, package name reflects the actual name of the bot.

WJunHong committed:
Change avatar images

Previously their appearance did not fit the personality of Chi.

Now, the avatar does fit Chi's cheery personality.

WJunHong committed:
Merge pull request #1 from WJunHong/branch-A-Assertions

Branch a assertions

WJunHong committed:
Merge branch 'master' into branch-A-CodeQuality

There were initial conflicts between branches.

Now there are no more conflicts.

WJunHong committed:
Merge pull request #2 from WJunHong/branch-A-CodeQuality

Branch a code quality

WJunHong committed:
Add goodbye message on bye

Previously, typing bye will simply terminate the program.

Now, typing bye will trigger good bye message, then terminate program.

WJunHong committed:
Finish help command

Previously, the help command was incomplete.

Now, the help command has features to assist users who are stuck on
error messages.

WJunHong committed:
Change DialogBox behaviour

Previously, DB will cut off long content like list command.

Now, the full content is shown.

WJunHong committed:
Add styles.css file, change help messages

Previously, styling of javaFX elements was done in the element itself.
Also, help message was not properly styled.

Now, styling is all done in 1 styles.css file, and help message is
better styled for readability.

Going forward, help message will have a hyperlink to user guide.
Also, error messages will be more helpful.

WJunHong committed:
Edit generic help method

Previously, help command message was not styled well.

Now, the styling is slightly more readable.

WJunHong committed:
Improve code quality of AddCommand class

Previously, there were some long methods in the class.

Now, the methods were SLAP-ed and broken down into further levels of
abstraction.

WJunHong committed:
Add more test cases

Previously, not many test cases were implemented.

Now, more test cases are added to test new parsing methods and tasklist
methods.

WJunHong committed:
Add hyperlink feature to help command

Previously, the generic help message had buggy UI.

Now, it the URL is separated from the initial message and there is a
working URL to the remote repository.

WJunHong committed:
Add more test cases and improve code quality

Previously, there were still some methods in add command not tested.

Now, testing has been made for them. Also, cleared data to prepare for
final screenshots.

WJunHong committed:
Set theme jekyll-theme-tactile

WJunHong committed:
Fix remaining kinks, add User Guide

Previously, the message for empty input was not getting printed,
and somehow tasks were not appended on the new line when stored in
data file.

Now, these changes are made. Also, user guide is added to help users
start using the app.

WJunHong committed:
Merge branch 'master' of https://github.com/WJunHong/ip

WJunHong committed:
Update README.md

Previously, README still uses Duke.

Now, full ChiBot README is available on GitHub

WJunHong committed:
Fix header

WJunHong committed:
Change error messages

Previously, messages did not tell users what they could do to correct
their input.

Now, they will be given a list of options to try out to rectify their
wrong input.

WJunHong committed:
Update README.md

Previously, user guide link was broken.

Now, it is available.

WJunHong committed:
Merge branch 'master' of https://github.com/WJunHong/ip

WJunHong committed:
Correct file name

Before, the user image name was incorrect, making the jar file unable
to run.

Now, it is updated and the jar file can run.

WJunHong committed:
Add jar file link

Before, the jar file link was missing in the project README.md.

Now, it is there.

WJunHong committed:
Handle file errors while parsing

Before, if the user were to change the data file, problems were not
detected or handled properly.

Now, badly formatted entries are ignored to prevent code break.

WJunHong committed:
Change delimiter check

Before, wrong regex was used.

Now, regex is corrected.

WJunHong committed:
Fix bug

Before, user input of 0 to delete and marks were not handled well.

Now, 0 is not treated as a valid index.
